testsuite/diem-backend/src/lib.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+use async_trait::async_trait;
+
+pub mod k8s_node;
+pub mod k8s_swarm;
+pub mod local_node;
+pub mod local_swarm;
+
+use diem_crypto::test_utils::KeyPair;
+use diem_crypto::ed25519::{Ed25519PrivateKey, Ed25519PublicKey};
+use diem_types::waypoint::Waypoint;
+use reqwest::{Client, Url};
+use diem_types::chain_id::ChainId;
+use std::fmt;
+use std::str::FromStr;
+use diem_client::Client as JsonRpcClient;
+
+#[derive(Clone)]
+enum InstanceBackend {
+    Swarm,
+}
+
+#[derive(Clone)]
+pub struct Cluster {
+    // guaranteed non-empty
+    validator_instances: Vec<Instance>,
+    fullnode_instances: Vec<Instance>,
+    lsr_instances: Vec<Instance>,
+    vault_instances: Vec<Instance>,
+    mint_key_pair: KeyPair<Ed25519PrivateKey, Ed25519PublicKey>,
+    waypoint: Option<Waypoint>,
+    pub chain_id: ChainId,
+}
+
+impl Cluster {
+    pub fn validator_instances(&self) -> &[Instance] {
+        &self.validator_instances
+    }
+
+    pub fn fullnode_instances(&self) -> &[Instance] {
+        &self.fullnode_instances
+    }
+
+    pub fn lsr_instances(&self) -> &[Instance] {
+        &self.lsr_instances
+    }
+
+    pub fn vault_instances(&self) -> &[Instance] {
+        &self.vault_instances
+    }
```

Comment:
My biggest question is what is the intent behind an `Instance`? Is it supposed to hold some information about a particular Node (ie address and json-rpc port, etc)? If its specific to validators then it shouldn't be reused fro things like LSR and Vault which might not have the same type of data (debug port, ac_port, etc)

Reply:
yea, this is the point I keeping thinking these days. Ideally my preference is we can have unified trait(interfaces) and also unified struct can hold information for both CT and Smoke and also AWS and Local. At first my thought is use current way in CT which some fields difference(what u asked above for LSR) can set to default or reserve as Option. But I found struct defined in smoke test does not have many common fields which means we may have more default and Option fields, but this obviously not the way I really like

Reply:
We have type safety for a reason, I'd lean more into having distinct types for these various players.